composer/datasets/cifar_hparams.py

```diff
@@ -162,12 +162,20 @@ def initialize_object(self, batch_size: int, dataloader_hparams: DataLoaderHpara
                     transforms.Normalize(cifar10_mean, cifar10_std),
                 ])
 
-            dataset = CIFAR10(
-                self.datadir,
-                train=self.is_train,
-                download=self.download,
-                transform=transformation,
-            )
+            def _get_dataset(download: bool):
+                return CIFAR10(
+                    self.datadir,
+                    train=self.is_train,
+                    download=download,
+                    transform=transformation,
+                )
+
+            if self.download:
+                if dist.get_local_rank() == 0:
+                    _get_dataset(download=True)
+                dist.barrier()
```

Comment:
This will fail if dist is not initialized. Suggestion, can we maybe use a contextmanager for this? something like:
```
    def __init__(download: bool):
        self.download = download
        self.wait_for_rank_zero = download and dist.initialized()
        
    def __enter__(self):
        if not self.wait_for_rank_zero: 
             return
        if dist.get_local_rank() == 0:
             self.download = True
         else:
             self.download = False
             dist.barrier()

def __exit__():
       if self.wait_for_rank_zero and dist.get_local_rank==0:
             dist.barrier()
           
```

so that this can be reused across datasets?

Reply:
`Dist` should be initialized when using the hparams codepath -- see https://github.com/mosaicml/composer/blob/81f3db523f6f7133262c5c4c8c53f7ef6060e1a2/composer/trainer/trainer_hparams.py#L424

Reply:
Also, see nit belong re: making this a lambda.